File: python/production/solver/solution_manager.py
from production.solver.solution import Solution
from production.solver.solution_node import Solution_node

class Solution_manager:

        def __init__(self):
            self.c_solutionList = list()
            for _ in range(0, 41):
                self.c_solutionList.append(list())

        def addSolution(self, p_rotationLinkedList, p_permutation, p_prevSolution,
                         p_value, p_floor):
            if len(self.c_solutionList[p_value]) < 40:
                self.c_solutionList[p_value].append(Solution_node(Solution(
                    p_rotationLinkedList.get_copy(), p_permutation.get_copy(), p_prevSolution)))

        def getBestUndeveloped(self):
            i = 40
            l_bestSolution = None
            while (i >= 0 and l_bestSolution == None):

                if len(self.c_solutionList[i]) > 0:
                    j = 0
                    while (j < len(self.c_solutionList[i]) and l_bestSolution == None):
                        l_node = self.c_solutionList[i][j]
                        j+=1
                        if not l_node.isDeveloped():
                            l_bestSolution = l_node.getSolution()
                            l_node.setDeveloped()
                i-=1
            return l_bestSolution

        def getBest(self):
            i = 40
            l_returnValue = None
            while (i >= 0 and l_returnValue == None):
                if len(self.c_solutionList[i]) > 0:
                    l_returnValue = self.c_solutionList[i][0].getSolution()
                i-=1
            return l_returnValue
        

        def getBestValue(self):
            i = 40
            l_returnValue = 0
            while (i >= 0 and l_returnValue == 0):
                if len(self.c_solutionList[i]) > 0:
                    l_returnValue = i
                i-=1
            return l_returnValue

File: csharp/Production/solver/SolutionManager.cs
using cube;
using System;
using System.Collections.Generic;


namespace solver
{
    public class SolutionManager
    {

        private List<List<SolutionNode>> c_solutionList = new List<List<SolutionNode>>();


        public SolutionManager()
        {
            int i;
            for (i = 0; i < 41; i++)
                c_solutionList.Add(new List<SolutionNode>());
        }

        public void addSolution(RotationSequence p_rotationLinkedList, Cube p_permutation, Solution p_prevSolution,
                         int p_value, int p_floor)
        {

            if (/*(p_value>=32 && getBestValue()>=36) ||*/ c_solutionList[p_value].Count < 40)
            {
                c_solutionList[p_value].Add(new SolutionNode(new Solution(p_rotationLinkedList.getCopy(), p_permutation.getCopy(), p_prevSolution)));

            }



        }


        public Solution getBestUndeveloped()
        {
            int i = 40;
            Solution l_bestSolution = null;
            while (i >= 0 && l_bestSolution == null)
            {
                if (c_solutionList[i].Count > 0)
                {
                    int j = 0;
                    while (j < c_solutionList[i].Count && l_bestSolution == null)
                    {
                        SolutionNode l_node = c_solutionList[i][j++];
                        if (!l_node.isDeveloped())
                        {
                            l_bestSolution = l_node.getSolution();
                            l_node.setDeveloped();
                        }
                    }
                }
                i--;
            }
            return l_bestSolution;
        }

        public Solution getBest()
        {
            int i = 40;
            Solution l_returnValue = null;
            while (i >= 0 && l_returnValue == null)
            {
                if (c_solutionList[i].Count > 0)
                    l_returnValue = c_solutionList[i][0].getSolution();
                i--;
            }
            return l_returnValue;
        }

        public int getBestValue()
        {
            int i = 40;
            int l_returnValue = 0;
            while (i >= 0 && l_returnValue == 0)
            {
                if (c_solutionList[i].Count > 0)
                    l_returnValue = i;
                i--;
            }
            return l_returnValue;
        }

        public class SolutionNode
        {
            private Solution c_solution;
            private Boolean c_isDeveloped;

            public SolutionNode(Solution p_solution)
            {
                c_isDeveloped = false;
                c_solution = p_solution;
            }

            public Solution getSolution()
            {
                return c_solution;
            }


            public Boolean isDeveloped()
            {
                return c_isDeveloped;
            }

            public void setDeveloped()
            {
                c_isDeveloped = true;
            }


        }
    }
}